src/fragmentarium/ui/info/GenreSelection.tsx

```diff
@@ -0,0 +1,203 @@
+import { Fragment, Genre } from 'fragmentarium/domain/fragment'
+import React, { useEffect, useRef, useState } from 'react'
+import {
+  Button,
+  Overlay,
+  OverlayTrigger,
+  Popover,
+  Tooltip,
+} from 'react-bootstrap'
+import Select from 'react-select'
+import classNames from 'classnames'
+import { usePrevious } from 'common/usePrevious'
+import withData from 'http/withData'
+import _ from 'lodash'
+
+type Props = {
+  fragment: Fragment
+  updateGenres: (genres: Genre[]) => void
+  genreOptions: readonly string[][]
+}
+
+function GenreSelection({
+  fragment,
+  updateGenres,
+  genreOptions,
+}: Props): JSX.Element {
+  const [selectedGenre, setSelectedGenre] = useState<Genre>()
+  const [genres, setGenres] = useState(fragment.genres)
+  const prevGenres = usePrevious(genres)
+  const [isSelectDisplayed, setIsSelectDisplayed] = useState(false)
+  const [
+    isDuplicateWarningDisplayed,
+    setIsDuplicateWarningDisplayed,
+  ] = useState(false)
+  const [isSelectedGenreUncertain, setIsSelectedGenreUncertain] = useState(
+    false
+  )
+  const target = useRef(null)
+
+  function handleChange(event) {
+    const newSelection: Genre = event.value
+    setSelectedGenre(newSelection)
+    if (!isGenreAlreadyPresent(newSelection)) {
+      setGenres((currentGenres) =>
+        _.sortBy([...currentGenres, newSelection], function (genre) {
+          return JSON.stringify(genreOptions).indexOf(
+            JSON.stringify(genre.category)
+          )
+        })
+      )
+    } else {
+      setIsDuplicateWarningDisplayed(true)
+    }
+  }
+  useEffect(() => {
+    if (JSON.stringify(prevGenres) !== JSON.stringify(genres)) {
+      updateGenres(genres)
+    }
+  })
```

Comment:
The update should be done in the event handler. Having a side effect on render is rather unexpected.

The result of the update is lost. If the `updateGenres` fails, the user sees the change, but nothing is saved.

Reply:
so there are three event handlers, selecting a genre, clicking uncertain checkbox and deleting a genre. All of those make change to the state `genres`. Everytime it changes there should be an API call. How exactly should i deal with that ? Do we have an example ?